test/mock/kernel/mock_framework.zig

```diff
@@ -0,0 +1,588 @@
+        ///
+        fn getFunctionType(comptime RetType: type, params: ...) type {
+            if (params.len == 0) return fn () RetType;
+            if (params.len == 1) return fn (@typeOf(params[0])) RetType;
+            if (params.len == 2) return fn (@typeOf(params[0]), @typeOf(params[1])) RetType;
+        }
+
+        ///
+        /// This tests a value passed to a function.
+        ///
+        /// Arguments:
+        ///     IN ExpectedType: type           - The expected type of the value to be tested.
+        ///     IN expected_value: ExpectedType - The expected value to be tested. This is what was
+        ///                                       passed to the functions.
+        ///     IN elem: DataElement            - The wrapped data element to test against the
+        ///                                       expected value.
+        ///
+        fn expectTest(comptime ExpectedType: type, expected_value: ExpectedType, elem: DataElement) void {
+            if (ExpectedType == void) {
+                // Can't test void as it has no value
+                warn("Can not test a value for void\n");
+                expect(false);
+            }
+
+            // Test that the types match
+            const expect_type = comptime getDataElementType(ExpectedType);
+            expectEqual(expect_type, DataElementType(elem));
+
+            // Types match, so can use the expected type to get the actual data
+            const actual_value = getDataValue(ExpectedType, elem);
+
+            // Test the values
+            expectEqual(expected_value, actual_value);
+        }
+
+        ///
+        /// This returns a value from the wrapped data element. This will be a test value to be
+        /// returned by a mocked function.
+        ///
+        /// Arguments:
+        ///     IN fun_name: []const u8         - The function name to be used to tell the user if
+        ///                                       there is no return value set up.
+        ///     IN/OUT action_list: *ActionList - The action list to extract the return value from.
+        ///     IN DataType: type               - The type of the return value.
+        ///
+        fn expectGetValue(comptime fun_name: []const u8, action_list: *ActionList, comptime DataType: type) DataType {
+            if (DataType == void) {
+                return;
+            }
+
+            if (action_list.*.popFirst()) |action_node| {
+                const action = action_node.data;
+                const expect_type = getDataElementType(DataType);
+
+                const ret = getDataValue(DataType, action.data);
+
+                expectEqual(DataElementType(action.data), expect_type);
+
+                // Free the node
+                action_list.*.destroyNode(action_node, GlobalAllocator);
+
+                return ret;
+            } else {
+                warn("No more test values for the return of function: " ++ fun_name ++ "\n");
+                expect(false);
```

Comment:
Are `expect(false)` and `unreachable` needed or can we just have one?

Reply:
`expect(false)` will call panic, therefore the code after is unreachable, but the compiler doesn't work this out, and as I return from this function, I need a return statement in each branch. This allows only a return for the branch that successes.

Reply:
:+1: